File: UXB/src/UXB/AbstractDevice.java
package UXB;

import UXB.Messages.Message;

import java.math.BigInteger;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Created by james on 9/12/16.
 *
 * Provides the basic functionality and requirements for devices using the USB
 */
public abstract class AbstractDevice<T extends AbstractDevice.Builder<T>> implements Device {

    private final Integer version;
    private final Optional<Integer> productCode;
    private final Optional<BigInteger> serialNumber;

    //list of the type of connectors attached to the device
    private final List<Connector.Type> connectorTypeList;
    private List<Connector> connectorList;

    protected AbstractDevice(Builder<T> builder) {
        this.productCode = builder.productCode;
        this.serialNumber = builder.serialNumber;
        this.version = builder.version;
        this.connectorTypeList = builder.connectorTypeList;

        connectorList = new ArrayList<>();

        //creating list of connectors to the device
        for(int i = 0; i < connectorTypeList.size(); i++) {
            Connector conn = new Connector(i, connectorTypeList.get(i), this);
            connectorList.add( conn );
        }
    }

    @Override
    public Optional<Integer> getProductCode() {
        return productCode;
    }

    @Override
    public Optional<BigInteger> getSerialNumber() {
        return serialNumber;
    }

    @Override
    public Integer getVersion() {
        return version;
    }

    @Override
    public DeviceClass getDeviceClass() {
        return null;
    }

    @Override
    public Integer getConnectorCount() {
        return connectorTypeList.size();
    }

    @Override
    // returns a list of connectors
    public List<Connector> getConnectors() {
        return new ArrayList<>(connectorList);
    }

    public void setConnectors(List<Connector> connectorList) {
        this.connectorList = new ArrayList<>(connectorList);
    }

    @Override
    // returns a new connector for the device with the same type as the connector of a given index
    public Connector getConnector(int index) {
        return connectorList.get(index);
    }

    //Checks the paramaters sent to a device's recv and throws any appropriate errors
    //Inputs: paramaters sent to recv
    //Does not output but rather throws errors if necessary
    protected void validateRecvArguments(Message message, Connector connector) {
        if(message == null || connector == null) {
            throw new NullPointerException("Recv cannot accept null arguments");
        }

        //checks if the peer is present
        //if it is, it then checks if the peer device is the right device
        if( !this.equals(connector.getDevice()) ) {
            throw new IllegalArgumentException("The connector sent to recv does not belong to this device");
        }
    }

    public Set<Device> peerDevices() {
        return connectorList.stream()
                .filter(connector -> connector.getPeer().isPresent()) //filters out any empty peers
                .map(connector -> connector.getPeer().get().getDevice()) //gets the peer device
                .collect(Collectors.toSet());
    }

    public Set<Device> reachableDevices() {
        Set<Device> emptyDeviceSet = new HashSet<>();
        return reachableDevicesHelper(emptyDeviceSet, this, null);
    }

    private Set<Device> reachableDevicesHelper(Set<Device> deviceSet, Device currDevice, Device deviceToFind){

        peerDeviceLoop:
        for (Device neighboringDevice : currDevice.peerDevices()) {

            if( neighboringDevice.equals(deviceToFind)) {
                deviceSet.add(neighboringDevice);
                break peerDeviceLoop;
            }

            if ( !deviceSet.contains(neighboringDevice) ) {
                deviceSet.add(neighboringDevice);
                reachableDevicesHelper(deviceSet, neighboringDevice, deviceToFind);
            }
        }
        return deviceSet;
    }

    public boolean isReachable(Device device) {
        Set<Device> emptyDeviceSet = new HashSet<>();
        return reachableDevicesHelper(emptyDeviceSet, this, device).contains(device);
    }


    public static abstract class Builder<T> {

        private Integer version;
        private Optional<Integer> productCode;
        private Optional<BigInteger> serialNumber;

        //list of the type of connectors attached to the device
        private List<Connector.Type> connectorTypeList;

        public Builder(Integer version){
            this.version = version;
            productCode = Optional.empty();
            serialNumber = Optional.empty();
            connectorTypeList = new ArrayList<>();
        }

        public T productCode (Integer productCode) {
            this.productCode = Optional.ofNullable(productCode);
            return getThis();
        }

        public T serialNumber (BigInteger serialNumber) {
            this.serialNumber = Optional.ofNullable(serialNumber);
            return getThis();
        }

        public T connectors (List<Connector.Type> connectors) {
            //initializing connector list with connectors to prevent copying reference
            this.connectorTypeList = connectors == null ?
                                    new ArrayList<>() :
                                    new ArrayList<>(connectors);
            return getThis();
        }

        protected abstract T getThis();

        protected List<Connector.Type> getConnectors() {
            return new ArrayList<>(connectorTypeList);
        }

        //Throws an error if the version is missing
        protected void validate() {
            if(version == null)
                throw new NullPointerException("A version was not given for this device.");
        }
    }
}

File: UXB/src/UXB/AbstractDevice.kt
package UXB

import UXB.Messages.Message

import java.math.BigInteger
import java.util.*
import java.util.stream.Collectors

/**
 * Created by james on 9/12/16.

 * Provides the basic functionality and requirements for devices using the USB
 */
abstract class AbstractDevice<T : AbstractDevice.Builder<T>> protected constructor(builder: AbstractDevice.Builder<T>) : Device {

    private val version: Int?
    private val productCode: Optional<Int>
    private val serialNumber: Optional<BigInteger>

    //list of the type of connectors attached to the device
    private val connectorTypeList: List<Connector.Type>
    private var connectorList: MutableList<Connector>? = null

    init {
        this.productCode = builder.productCode
        this.serialNumber = builder.serialNumber
        this.version = builder.version
        this.connectorTypeList = builder.connectorTypeList

        connectorList = ArrayList<Connector>()

        //creating list of connectors to the device
        for (i in connectorTypeList.indices) {
            val conn = Connector(i, connectorTypeList[i], this)
            connectorList!!.add(conn)
        }
    }

    override fun getProductCode(): Optional<Int> {
        return productCode
    }

    override fun getSerialNumber(): Optional<BigInteger> {
        return serialNumber
    }

    override fun getVersion(): Int? {
        return version
    }

    override fun getDeviceClass(): DeviceClass? {
        return null
    }

    override fun getConnectorCount(): Int? {
        return connectorTypeList.size
    }

    override // returns a list of connectors
    fun getConnectors(): List<Connector> {
        return ArrayList(connectorList!!)
    }

    fun setConnectors(connectorList: List<Connector>) {
        this.connectorList = ArrayList(connectorList)
    }

    override // returns a new connector for the device with the same type as the connector of a given index
    fun getConnector(index: Int): Connector {
        return connectorList!![index]
    }

    //Checks the paramaters sent to a device's recv and throws any appropriate errors
    //Inputs: paramaters sent to recv
    //Does not output but rather throws errors if necessary
    protected fun validateRecvArguments(message: Message?, connector: Connector?) {
        if (message == null || connector == null) {
            throw NullPointerException("Recv cannot accept null arguments")
        }

        //checks if the peer is present
        //if it is, it then checks if the peer device is the right device
        if (this != connector.device) {
            throw IllegalArgumentException("The connector sent to recv does not belong to this device")
        }
    }

    //forwards message to all peer connectors, barring incomingConnector
    //if incomingConnector is null, this should send the message to all peers
    //Allowing message to be null as the error will be picked up by validateRecvArguments
    protected fun sendMessageToAllPeersExceptIncomingConnector(message: Message, incomingConnector: Connector) {
        this.connectors.filter({ conn -> conn.getPeer().isPresent() && conn != incomingConnector }).map({ connsWithPeers -> connsWithPeers.getPeer().get() }) // grab the peer connectors
                .forEach({ peerConns -> peerConns.recv(message) })
    }

    override fun peerDevices(): Set<Device> {
        return connectorList!!.stream().filter({ connector -> connector.getPeer().isPresent() }) //filters out any empty peers
                .map({ connector -> connector.getPeer().get().getDevice() }) //gets the peer device
                .collect(Collectors.toSet<Device>())
    }

    override fun reachableDevices(): Set<Device> {
        val emptyDeviceSet = HashSet<Device>()
        return reachableDevicesHelper(emptyDeviceSet, this, null)
    }

    private fun reachableDevicesHelper(deviceSet: MutableSet<Device>, currDevice: Device, deviceToFind: Device?): Set<Device> {

        peerDeviceLoop@ for (neighboringDevice in currDevice.peerDevices()) {

            if (neighboringDevice == deviceToFind) {
                deviceSet.add(neighboringDevice)
                break@peerDeviceLoop
            }

            if (!deviceSet.contains(neighboringDevice)) {
                deviceSet.add(neighboringDevice)
                reachableDevicesHelper(deviceSet, neighboringDevice, deviceToFind)
            }
        }
        return deviceSet
    }

    override fun isReachable(device: Device): Boolean {
        val emptyDeviceSet = HashSet<Device>()
        return reachableDevicesHelper(emptyDeviceSet, this, device).contains(device)
    }


    abstract class Builder<T>(private val version: Int?) {
        private var productCode: Optional<Int>? = null
        private var serialNumber: Optional<BigInteger>? = null

        //list of the type of connectors attached to the device
        private var connectorTypeList: List<Connector.Type>? = null

        init {
            productCode = Optional.empty<Int>()
            serialNumber = Optional.empty<BigInteger>()
            connectorTypeList = ArrayList<Type>()
        }

        fun productCode(productCode: Int?): T {
            this.productCode = Optional.ofNullable<Int>(productCode)
            return `this`
        }

        fun serialNumber(serialNumber: BigInteger): T {
            this.serialNumber = Optional.ofNullable(serialNumber)
            return `this`
        }

        fun connectors(connectors: List<Connector.Type>?): T {
            //initializing connector list with connectors to prevent copying reference
            this.connectorTypeList = if (connectors == null)
                ArrayList<Type>()
            else
                ArrayList(connectors)
            return `this`
        }

        protected abstract val `this`: T

        protected val connectors: List<Connector.Type>
            get() = ArrayList(connectorTypeList!!)

        //Throws an error if the version is missing
        protected open fun validate() {
            if (version == null)
                throw NullPointerException("A version was not given for this device.")
        }
    }
}
